### ISSUE TITLE
Weird error with reduce on empty vector (or matrix)

### ISSUE THREAD

**Issue by eriknw** [jim22k/grblas#8]
```python-traceback
In [1]: import grblas

In [2]: grblas.init()

In [3]: x = grblas.Vector.new_from_type(grblas.dtypes.INT64)

In [4]: x.reduce(grblas.binary.plus).new()
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-3-a93c2ce4a2a7> in <module>
      1 x = grblas.Vector.new_from_type(grblas.dtypes.INT64)
----> 2 x.reduce(grblas.binary.plus).new()

~/Hive/git/grblas/grblas/base.py in new(self, dtype, mask)
    254             output = self.output_constructor()
    255         if mask is None:
--> 256             output.update(self)
    257         else:
    258             if not isinstance(mask, output.__class__):

~/Hive/git/grblas/grblas/base.py in update(self, delayed)
    117         Convenience function when no output arguments (mask, accum, replace) are used
    118         """
--> 119         return self._update(delayed)
    120
    121     def _update(self, delayed, mask=NULL, accum=NULL, replace=False):

~/Hive/git/grblas/grblas/base.py in _update(self, delayed, mask, accum, replace)
    187
    188         # Make the GraphBLAS call
--> 189         check_status(delayed.func(*call_args))
    190
    191     def _extract_element(self, resolved_indexes):

TypeError: initializer for ctype 'struct GB_Monoid_opaque *' must be a pointer to same type, not cdata 'struct GB_BinaryOp_opaque *'
```
Using a monoid is fine:
```python
In [5]: x.reduce(grblas.monoid.plus).new().value
Out[5]: 0
```

**Comment by eriknw** [comment]
TBH, I don't know what expected behavior is here.  Typically, we rely on the monoid to get the identity, which makes such things as vacuous truth "just work".

**Comment by eriknw** [comment]
Okay, it appears that reductions to scalars *must* use monoids.  It may be helpful for us improve some of the error messages.

Reductions from Matrix to Vector can use either a BinaryOp or a Monoid.

According to the doc, SuiteSparse:GraphBLAS appears to extend the spec by allowing a Matrix to Scalar reduction using a BinaryOp.  I don't see how to do this from Python though.

tl;dr: reducing to scalar must use a monoid.  If there are no values, then the identity from the monoid is used.

Closing.
